### PR TITLE
Add tpm_check_buffer_size test case

### PR DESCRIPTION
**Pull request by qcheng-redhat**
Add tpm_check_buffer_size test case

ID: 2189100

Signed-off-by: qcheng-redhat <qcheng@redhat.com>

### CODE REVIEW

**Comment by qcheng-redhat** [comment]
RHEL 8:
 (1/1) Host_RHEL.m8.u9.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35:PASS (6.20 s)

RHEL 9:
 (1/1) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35:PASS (11.67 s)

**Comment by qcheng-redhat** [comment]
Hi @PaulYuuu @XueqiangWei ,  please help review this patch. Thanks.

File: qemu/tests/tpm_check_buffer_size.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    1. Download and install libtpms-devel rpm
+    2. Clone libtpms test tree
+    3. Comple and execute test cases
+
+    :param test: QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+
+    s, o = process.getstatusoutput("rpm -q libtpms-devel", shell=True)
```
Comment (PaulYuuu) [inline qemu/tests/tpm_check_buffer_size.py:25]:
```suggestion
    s = process.system("rpm -q libtpms-devel", shell=True)
```
Or utils_packages.package_install to try to install it.

File: qemu/tests/tpm_check_buffer_size.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    if s:
+        try:
+            libtpms_rpm = process.system_output("rpm -q libtpms", shell=True)
```
Comment (PaulYuuu) [inline qemu/tests/tpm_check_buffer_size.py:29]:
```suggestion
            libtpms_rpm = process.getoutput("rpm -q libtpms", shell=True)
```
`system_output` returns byte-like, `getoutput` returns string.

File: qemu/tests/tpm_check_buffer_size.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            libtpms_devel_rpm = (
+                libtpms_rpm.split(b'-', 1)[0] +
+                b'-devel-' +
+                libtpms_rpm.split(b'-', 1)[1]).decode()
+            process.system(
+                "brew download-build --rpm %s" %
+                libtpms_devel_rpm, shell=True)
+            libtpms_devel_rpm = libtpms_devel_rpm + '.rpm'
+            process.system("rpm -i %s" % libtpms_devel_rpm, shell=True)
```
Comment (PaulYuuu) [inline qemu/tests/tpm_check_buffer_size.py:38]: Better to remove the rpm in the finish.

File: qemu/tests/cfg/tpm_check_buffer_size.cfg [hunk 4]
```diff
@@ -0,0 +1,8 @@
+- tpm_check_buffer_size:
+  virt_test_type = qemu
+  type = tpm_check_buffer_size
+  start_vm = no
+  repo_url = https://github.com/stefanberger/libtpms
+  test_case = tpm2_cve-2023-1017.c;
+  test_case += tpm2_cve-2023-1018.c
+  build_execute_cmd = gcc tests/%s -ltpms && ./a.out
```
Comment (PaulYuuu) [inline qemu/tests/cfg/tpm_check_buffer_size.cfg:8]: Consider using `-o tpm2_cve-2023-1017/tpm2_cve-2023-1018` to replace `a.out`?

File: qemu/tests/tpm_check_buffer_size.py (hunk 1 above)
Comment (qcheng-redhat) [reply qemu/tests/tpm_check_buffer_size.py]: Updated.

File: qemu/tests/tpm_check_buffer_size.py (hunk 3 above)
Comment (qcheng-redhat) [reply qemu/tests/tpm_check_buffer_size.py]: Add remove step in the finally section .

File: qemu/tests/cfg/tpm_check_buffer_size.cfg (hunk 4 above)
Comment (qcheng-redhat) [reply qemu/tests/cfg/tpm_check_buffer_size.cfg]: Updated.

File: qemu/tests/cfg/tpm_check_buffer_size.cfg [hunk 5]
```diff
@@ -0,0 +1,8 @@
+- tpm_check_buffer_size:
+  virt_test_type = qemu
```
Comment (PaulYuuu) [inline qemu/tests/cfg/tpm_check_buffer_size.cfg:2]: 4x spaces indentation please.

Comment (qcheng-redhat) [reply qemu/tests/cfg/tpm_check_buffer_size.cfg]: Updated. Thanks

File: qemu/tests/tpm_check_buffer_size.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            "Build and execute test cases", test.log.info)
+        for test_case in params.get("test_case").split(";"):
+            test_case_o = test_case.split(".")[0]
+            build_execute_cmd = params["build_execute_cmd"] % (
+                test_case, test_case_o, test_case_o)
+            process.system("cd %s && %s" %
+                           (repo_dir, build_execute_cmd), shell=True)
+    finally:
+        rmtree(repo_dir, ignore_errors=True)
+        process.system("rm -f /home/%s" % libtpms_devel_rpm,
```
Comment (PaulYuuu) [inline qemu/tests/tpm_check_buffer_size.py:55]: When the system already installed the libtpms-devel pkg, this operation will fail.
```
2023-04-26 21:37:18,238 avocado.utils.process INFO | Running 'rm -f /home/libtpms-devel-0.9.1-3.20211126git1ff6fe1f43.el9_2.aarch64'
2023-04-26 21:37:18,289 avocado.utils.process DEBUG| [stderr] shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
2023-04-26 21:37:18,289 avocado.utils.process INFO | Command 'rm -f /home/libtpms-devel-0.9.1-3.20211126git1ff6fe1f43.el9_2.aarch64' finished with 0 after 0.002187907s
```

And I don't think we need the next line(`rpm -e`), just check if the rpm was installed, if not, then install it.

Comment (qcheng-redhat) [reply qemu/tests/tpm_check_buffer_size.py]: Updated the directory issue.  Thanks.

Comment (qcheng-redhat) [reply qemu/tests/tpm_check_buffer_size.py]: Updated the directory issue.  Thanks.

File: qemu/tests/tpm_check_buffer_size.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    3. Comple and execute test cases
+
+    :param test: QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+
+    libtpms_rpm = process.getoutput("rpm -q libtpms", shell=True)
+    libtpms_devel_rpm = (libtpms_rpm.split("-", 1)[0] +
+                         "-devel-" + libtpms_rpm.split("-", 1)[1])
```
Comment (PaulYuuu) [inline qemu/tests/tpm_check_buffer_size.py:27]:
```suggestion
    libtpms_rpm = process.getoutput("rpm -q libtpms", shell=True).strip()
    libtpms_devel_rpm = (libtpms_rpm.replace("libtpms", "libtpms-devel")
```

Comment (qcheng-redhat) [reply qemu/tests/tpm_check_buffer_size.py]: Updated.

**Review by PaulYuuu — APPROVED** [review]

**Comment by XueqiangWei** [comment]
Tested it with virtio_blk and virtio_scsi, the results were passed.

 (1/4) Host_RHEL.m9.u3.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: STARTED
 (1/4) Host_RHEL.m9.u3.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: PASS (12.11 s)
 (2/4) Host_RHEL.m9.u3.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: STARTED
 (2/4) Host_RHEL.m9.u3.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: PASS (6.64 s)
 (3/4) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: STARTED
 (3/4) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: PASS (6.31 s)
 (4/4) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: STARTED
 (4/4) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: PASS (7.63 s)

**Comment by XueqiangWei** [comment]
LGTM,
Acked-by: xuwei@redhat.com

**Review by XueqiangWei — APPROVED** [review]

**Pull request merged**